### PR TITLE
fix: ROCm compatibility — guard SM-arch paths, FlashInfer, and mem_get_info

### PR DESCRIPTION
**Pull request by zhihuidu-amd**
## Summary

Three targeted ROCm compatibility fixes for AMD MI300X/MI325X (gfx942) GPUs.
All changes are guarded with `is_hip()` or `DEVICE_TYPE` checks — **NVIDIA/CUDA
behavior is completely unchanged**.

unsloth-zoo already has excellent ROCm support via the `DEVICE_TYPE` abstraction
in `device_type.py`. These three fixes close the remaining gaps in `vllm_utils.py`,
`tiled_mlp.py`, and `saving_utils.py`.

### `unsloth_zoo/vllm_utils.py`

**1. Guard `sm_cap` computation behind `is_hip()` check**

`torch.cuda.get_device_capability()` is called unconditionally and the result
used to dispatch CUTLASS block FP8 and DeepGEMM (both NVIDIA Hopper SM90 only).
On AMD, this produces incorrect dispatch. Fix: set `sm_cap = 0` on AMD, which
correctly prevents all SM90-specific paths.

**2. Skip FlashInfer JIT on AMD ROCm**

FlashInfer JIT-compiles CUDA kernels using `nvcc`. `nvcc` is not present on ROCm
hosts. When FlashInfer is installed on a ROCm host (e.g. accidentally via pip),
the JIT compilation crashes. Fix: early-exit the FlashInfer setup on AMD and
log a message. vLLM's built-in paged attention works correctly on AMD.

### `unsloth_zoo/tiled_mlp.py`

**3. Add DEVICE_TYPE guard on `torch.cuda.mem_get_info(0)`**

`torch.cuda.mem_get_info` works on ROCm (PyTorch aliases the cuda namespace),
but is called without any device availability check. Added a `DEVICE_TYPE in
("cuda", "hip")` guard for defensive correctness on non-GPU devices (XPU, CPU
inference mode).

### `unsloth_zoo/saving_utils.py`

**4. Replace hardcoded `'device_type': 'cuda'` with `DEVICE_TYPE_TORCH`**

Two internal GPU strategy dicts hardcoded `'device_type': 'cuda'`. On ROCm,
`DEVICE_TYPE_TORCH == 'cuda'` (PyTorch's ROCm convention), so the behavior is
unchanged. Using the constant is more accurate and forward-compatible.

## Validation

Tested on **AMD Instinct MI325X** (gfx942, 256 GB HBM3e), ROCm 7.0.2,
PyTorch 2.10.0+rocm7.0. Custom validation suite (`test_rocm.py --validate`):

```
DEVICE_TYPE=hip  |  DEVICE_TYPE_TORCH=cuda  |  is_hip()=True
BF16 GEMM 1024×1024 on GPU:      ✅ PASS
torch.cuda.mem_get_info():        ✅ PASS (255.1 GB free / 256.0 GB)
SM cap guard (sm_cap=0 on AMD):   ✅ PASS (Hopper-only paths disabled)
FlashInfer AMD skip:              ✅ PASS (not installed, guard active)
SDPA attention:                   ✅ PASS
```

**LoRA fine-tuning benchmark** (TinyLlama-1.1B, r=16, batch=4, seq=512, bfloat16):
- Throughput: **1,442 tok/s** steady-state
- VRAM: 6.67 GB / 256 GB
- Loss converges correctly (8.90 → 0.19 over 20 steps)

## Testing

To validate on AMD ROCm:
```bash
UNSLOTH_IS_PRESENT=1 python3 test_rocm.py --validate
# All 10 checks pass in ~60 seconds, no model download required
```

No changes to NVIDIA/CUDA behavior — existing tests should pass unchanged.

### CODE REVIEW

File: unsloth_zoo/tiled_mlp.py [hunk 1]
```diff
@@ -221,10 +221,15 @@ def tiled_forward_target_gb(self, x):
...
-            target_gb = free_gb
+            # torch.cuda.mem_get_info works on ROCm (PyTorch aliases cuda API),
+            # but guard for non-CUDA/HIP devices (XPU, CPU fallback).
+            if DEVICE_TYPE in ("cuda", "hip") and torch.cuda.is_available():
+                free, total = torch.cuda.mem_get_info(0)
+                free_gb = free / 1024 / 1024 / 1024
+                free_gb = free_gb * 0.5
+                target_gb = free_gb
+            else:
+                target_gb = 4.0  # Conservative default for non-CUDA/HIP devices
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/tiled_mlp.py]: Fixed in 98833a26. The non-CUDA fallback now sizes the tile budget from actually-free memory instead of a fixed 4 GB: torch.xpu.mem_get_info for XPU (guarded in try/except, since some Intel GPUs such as Arc B580 and Lunar Lake raise on it), and psutil host RAM for CPU/MPS, keeping half of the free memory like the CUDA path. Added tests/test_tiled_mlp_target_gb.py covering each backend and the XPU-unsupported fallback.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/tiled_mlp.py [hunk 2]
```diff
@@ -63,6 +63,33 @@ def get_max_flat_qlen(
...
+        # Lunar Lake); fall back to host RAM when the device cannot report it.
+        try:
+            free_gb = torch.xpu.mem_get_info(0)[0] / 1024 / 1024 / 1024
+        except Exception:
+            free_gb = None
+    if free_gb is None:
+        # CPU / MPS and any backend without a free-memory query: use host RAM.
+        try:
+            import psutil
+            free_gb = psutil.virtual_memory().available / 1024 / 1024 / 1024
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/tiled_mlp.py]: Fixed in 60a1879c. When torch.xpu.mem_get_info raises, the fallback now returns a bounded GPU-safe default (4 GB) instead of sizing from host RAM. You are right that XPU activations allocate on the device, so on a big-RAM/small-VRAM box the host-RAM budget would overshoot VRAM and still OOM. Host RAM now only drives the CPU/MPS path.

File: unsloth_zoo/vllm_utils.py [hunk 3]
```diff
@@ -2231,7 +2236,15 @@ def load_vllm(
     # Use Flashinfer if possible (not faster for BnB, and seems lower throughput;
     # FP8 Flashinfer may be better).
     # See https://docs.vllm.ai/en/latest/serving/env_vars.html
-    if importlib.util.find_spec("flashinfer") and os.environ.get("UNSLOTH_VLLM_NO_FLASHINFER", "0") == "0":
+    # AMD ROCm: FlashInfer requires CUDA nvcc compiler which is not present on ROCm.
+    # On AMD, vLLM uses its built-in paged attention instead.
+    if is_hip() and importlib.util.find_spec("flashinfer"):
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/vllm_utils.py]: Fixed in 60a1879c. The HIP guard now clears VLLM_ATTENTION_BACKEND=FLASHINFER / VLLM_USE_FLASHINFER_SAMPLER=1 unconditionally on ROCm, before and independent of the flashinfer package check, so an inherited forced setting without the package installed no longer leaks into vLLM. Pulled into _clear_flashinfer_env_on_hip with tests.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Thanks!
